Signup: Domain step should show empty state when the query is emptied

alshakero opened the pull request:
#### Changes proposed in this Pull Request

Fresh start (no results) => clearing the search query while suggestions are loading renders empty state
Existing results => clearing renders previous returned domain suggestions

#### Testing instructions

1. Go to [/start/domains](https://container-ecstatic-hoover.calypso.live/start) and allow 3rd party cookies.
2. Start typing to search for a domain and while the results are still loading, clear the search query input
3. **UI should go back to empty state.**
4. If you cleared the input, type one character, **search shouldn't start and the empty state should prevail**.
5. Type in 2+ characters, search **should start and results should be shown as you'd expect**.

Demo: https://user-images.githubusercontent.com/14192054/124305375-af115900-db6d-11eb-99c6-b7d71bd966a9.mov

Fixes: https://github.com/Automattic/wp-calypso/issues/53752

alshakero committed:
Show empty state when the query is deleted

razvanpapadopol reviewed (dismissed):
I can confirm that the ever-loading bug reported in #53752 is now fixed 🎉 

However, there's a regression that can be reproduced if you clear the search box after the results finished loading: "show more" button is there and you can click it twice before it disappears.

https://user-images.githubusercontent.com/14192054/123835056-c30c4f00-d910-11eb-8bc3-2897722968ee.mov

Also, there is one more case when a filter is applied and, probably because of very few results, there is a `register-domain-step__filter-reset-notice` button displayed. This button is still displayed even after the search box is cleared. Here, an option could be to reset also the filters when clearing the search box? 
<img width="1046" alt="Screenshot 2021-06-29 at 19 42 17" src="https://user-images.githubusercontent.com/14192054/123836496-71fd5a80-d912-11eb-869b-da1f5262f8ea.png">

alshakero committed:
Handle empty query in every render function

alshakero commented:
I addressed your feedback, Razvan. I handled the empty query in every render function. 

>  Here, an option could be to reset also the filters when clearing the search box?

I feel that would be too magical.

razvanpapadopol commented:
> > Here, an option could be to reset also the filters when clearing the search box?
> 
> I feel that would be too magical.

Probably yes 😅 
Let's leave this puzzle for @ollierozdarz and @poligilad-auto to solve separately when re-thinking TLD filters.

razvanpapadopol commented:
> I addressed your feedback, Razvan. I handled the empty query in every render function.

To avoid repetition, isn't possible to do a conditional rendering directly in the main `render` method?
```
{ this.state.lastQuery?.length && ( <> [[All the 'render...' calls here]] </> ) }
```

Also, there are still these notices left. They are not included in any of the 'render...' methods: https://github.com/Automattic/wp-calypso/blob/5ad06d1cb829a16a9d64732d49f0bf58bda771a8/client/components/domains/register-domain-step/index.jsx#L474-L489

For example, I can get one like this:

<img width="600" alt="Screenshot 2021-06-30 at 14 51 19" src="https://user-images.githubusercontent.com/14192054/123956681-a117d880-d9b3-11eb-89b2-24b33fa42bd8.png">
<img width="600" alt="Screenshot 2021-06-30 at 14 51 36" src="https://user-images.githubusercontent.com/14192054/123956688-a37a3280-d9b3-11eb-87b5-1b3617d141a6.png">

alshakero committed:
Restructure rendering

alshakero commented:
> To avoid repetition, isn't possible to do a conditional rendering directly in the main render method?

@razvanpapadopol Great advice. I followed it in the last commit.

razvanpapadopol commented:
After the last commit, e2e tests ([desktop](https://user-images.githubusercontent.com/14192054/124091120-29eb4e80-da5e-11eb-840f-c31df25b524f.png) and [mobile](https://user-images.githubusercontent.com/14192054/124091096-25269a80-da5e-11eb-959c-fb3fa7ecffad.png)) started to fail because of the launch flow. There, the "empty state" should look like this:
<img width="1281" alt="Screenshot 2021-07-01 at 11 19 46" src="https://user-images.githubusercontent.com/14192054/124091294-599a5680-da5e-11eb-876d-68b458eed514.png">
- 2 suggestions
- Use a domain I own
- Skip

Then, when searching, more suggestions are rendered

alshakero committed:
Fix the empty state in an entirely differnet way

razvanpapadopol commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -764,11 +764,7 @@ class RegisterDomainStep extends React.Component {
 		const loadingResults = Boolean( cleanedQuery );
-		const isInitialQueryActive = searchQuery === this.props.suggestion;
-
-		if ( isEmpty( cleanedQuery ) ) {
-			return;
-		}
```
Choosing to continue to update the state at this point (no valid query) is indeed going to clear the results and prevent getting stuck in loading mode. But unfortunately, this will undo what @klimeryk was trying to improve in https://github.com/Automattic/wp-calypso/pull/39664 for `/domains/add`.

Also, the bug we're trying to fix seems to be reproducing on that page as well so I've updated our new issue https://github.com/Automattic/wp-calypso/issues/53752 to mention it.

Then, I've updated the A/C: 

> - [ ] When the search query is cleared in any way, show empty state for the domain results; 

to 

> - [ ] When the search query is cleared (or reduced to 1 character), show the previous domain results or empty state if there aren't any;

razvanpapadopol replied:
With https://github.com/Automattic/wp-calypso/pull/54124/commits/57e758ff5b3e11188369b60ac9c7219c4f090f2d we don't reset the suggestions when the search query is changed. The loading state will make sure they remain hidden anyway. 

Fixed behavior looks like this:
https://user-images.githubusercontent.com/14192054/124305375-af115900-db6d-11eb-99c6-b7d71bd966a9.mov

* No results => clearing the search query while suggestions loading renders empty state
* Existing results => clearing renders previous results

Commit:
Stop setting the domain suggestion results to `null` when changing search query.
